### PR TITLE
Use getpwuid(3) in corehost for the $HOME-less user

### PR DESCRIPTION
**Pull request by am11**
In other parts of dotnet/runtime repo <sup>[[1](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Native/Unix/System.Native/pal_networking.c#L3038-L3056)] [[2](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/IO/PersistedFiles.Unix.cs#L125-L137)] [[3](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Native/Unix/System.Native/pal_uid.c#L65-L68)]</sup>, we attempt to retrieve user home directory from `getpwuid_r` when the environment variable `HOME` is not set. Use the same fallback in corehost.

### CODE REVIEW

**Comment by vitek-karas** [comment]
I have to defer to @janvorli on this. It looks good, but:
* we want the directory to be only accessible by the current user - paths accessible by other (non-root) users are potentially problematic as other users could DoS the extraction by pre-creating the right files/directories.
* if no such directory is available, I think it's better to fail and ask users to specify the extraction directory explicitly

In short under no condition should we extract to potentially DoS sensitive location by default.

If the `getpwuid` offers such guarantees, I'd be happy to use that as it solves real issues. But if these guarantees are not there, I think it would be better to let it fail. (I don't know if the other usages in the runtime may have similar requirements or not).

**Comment by janvorli** [comment]
@vitek-karas the getpwuid gets a record for the specified user (in the change, it is the current user) from the OS user database, the /etc/passwd file. That database, besides user names also contains the user home directory. The HOME env variable, if present, comes from the same source. It is set by login, ssh or a graphics login manager.

**Comment by am11** [comment]
> we want the directory to be only accessible by the current user - paths accessible by other (non-root) users are potentially problematic as other users could DoS the extraction by pre-creating the right files/directories.

Based on my understanding of this, the `else` case being modified here:

1. has at least same level of guarantee that its pairing `if` block is providing, where we read `$HOME` environment variable.
2. is more or at least as "reliable" than the `if` case, since getpwuid(3) reads information from system user database (same database which holds credentials that let user log in to the system).

**Comment by vitek-karas** [comment]
Thanks a lot for confirmation!

I think the only missing piece is to add a test. Something like - remember `$HOME`, run an extraction with `$HOME` set to empty, see if it extracted to the location matches the remembered location. Similar to the tests here: https://github.com/dotnet/runtime/blob/main/src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/BundleExtractToSpecificPath.cs

**Comment by am11** [comment]
Sure, I am working on a test case. However, `BundleExtractToSpecificPath` are not currently running in the CI. Installer legs are green, but e.g. https://github.com/dotnet/runtime/blob/a56b7321692d682bedf328c53dc8d47bbb193154/src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/BundleExtractToSpecificPath.cs#L233 didn't fail. Note that I changed the error message such that there is no period after `is not defined` anymore; due to this condition:
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/installer/tests/Directory.Build.props#L14-L15
`BundleExtractToSpecificPath` is marked with the trait `FlakyAppHostTests` that only runs on `runtime-staging` pipeline, and staging doesn't seem to have any installer leg.

I will, however, run them locally on macOS (11.5.2) and Linux (Ubuntu 20.04) before pushing the test, assuming that'd be enough. :)

**Comment by vitek-karas** [comment]
Hmm - @agocke has been looking into the "flakiness" here https://github.com/dotnet/runtime/issues/53587. Last I heard it's basically a race condition in the OS and we're looking into potential workarounds. I wasn't aware of the fact that we actually disabled these tests in the CI. They should be at the very least enabled everywhere but on Linux.

File: src/native/corehost/hostmisc/pal.unix.cpp [hunk 1]
```diff
@@ -352,7 +352,30 @@ bool get_extraction_base_parent_directory(pal::string_t& directory)
...
+        if (pwuid != NULL)
+        {
+            directory.assign(pwuid->pw_dir);
+            if (is_read_write_able_directory(directory))
+            {
+                return true;
+            }
+            else
+            {
+                trace::error(_X("Failed to determine default extraction location. Environment variable '$HOME' is not defined and directory reported by getpwuid(3) [%s] either doesn't exist or is not accessible for read/write."), pwuid->pw_dir);
```
Comment (janvorli) [inline src/native/corehost/hostmisc/pal.unix.cpp:372]: A nit - does it have any value to mention getpwuid(3) instead of just getpwuid?

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

File: src/native/corehost/hostmisc/pal.unix.cpp (hunk 1 above)
Comment (am11) [reply src/native/corehost/hostmisc/pal.unix.cpp:372]: `man man` notes:
```
MANUAL SECTIONS
       The standard sections of the manual include:

       1      User Commands
       2      System Calls
       3      C Library Functions
       4      Devices and Special Files
       5      File Formats and Conventions
       6      Games et. Al.
       7      Miscellanea
       8      System Administration tools and Deamons
```
`getpwuid(3)` indicates to the user that it is a system C library function (rather than something of off .NET codebase). I can delete `3` if that's preferred. :)

Comment (janvorli) [reply src/native/corehost/hostmisc/pal.unix.cpp:372]: I would prefer not to have it there.

**Review by vitek-karas — APPROVED** [review]
Thanks a lot!

**Comment by vitek-karas** [comment]
I think we should wait a couple of days. The CI is under water due to the high load right before RC2 cut-off (and there are also some issues causing more trouble). Since this is for 7 only, it's OK if we wait on this a bit and rebase-rerun and merge then.

**Comment by am11** [comment]
I ran all bundle tests locally and they passed (works on my machine ™️)

> it's OK if we wait on this a bit and rebase-rerun and merge then.

TYT, there is no rush but TBH, blocking this PR on the whole "bundle test flakiness", which has been going on since before .NET 5 release, is an overkill. There is no PR on horizon which is fixing/removing the `FlakyAppHostTests` trait. I was considering this PR as a minor improvement over your patch merged just yesterday https://github.com/dotnet/runtime/commit/1517985a85ed8af90e762f5b9ebeefd5827096e3, both patches are touching/improving same test methods.

**Comment by vitek-karas** [comment]
Sorry for not being clear. I didn't mean that we should hold this on the test flakiness. My comment about holding this PR for a bit was solely to not stress CI even more than it already is today (and also hopefully we'll see less infra issues after the RC 2 cutoff).

**Comment by vitek-karas** [comment]
Thanks a lot @am11!
